vispy/visuals/bar.py

```diff
@@ -0,0 +1,152 @@
+                        raise ValueError('bottom should only have one column')
+
+                else:
+                    raise ValueError('3d is not expected, just 2d')
+
+                if bottom[0] != height[0]:
+                    raise ValueError('bottom and height have to have the same length')
+        else:
+            if len(bottom) != len(height):
+                raise ValueError('bottom and height have to have the same length')
+            else:
+                index = np.arange(len(bottom))
+
+    if isinstance(data, np.ndarray):
+        if len(data.shape) == 1:
+            height = data.reshape(data.shape[0], 1)
+            index = np.arange(height.shape[0])
+            bottom = np.zeros(height.shape)
+
+        elif len(data.shape) == 2:
+            if data.shape[1] == 1:
+                height = data
+                index = np.arange(height.shape[0])
+                bottom = np.zeros(height.shape)
+
+            elif data.shape[1] == 2:
+                tmp_height = data[:, 0]
+                height = tmp_height.reshape(tmp_height.shape[0], 1)
+                index = np.arange(height.shape[0])
+                tmp_bottom = data[:, 1]
+                bottom = tmp_bottom.reshape(tmp_bottom.shape[0], 1)
+
+            else:
+                raise ValueError('data should have one column for height and an optionally a second for bottom of '
+                                 'bars')
+
+        else:
+            raise ValueError('3d is not expected, just 2d')
+
+    elif height is not None:
+        if isinstance(height, np.ndarray):
+            if len(height.shape) == 1:
+                height = height.reshape(height.shape[0], 1)
+                index = np.arange(height.shape[0])
+                if bottom is None:
+                    bottom = np.zeros(height.shape)
+
+            elif len(height.shape) == 2:
+                if height.shape[1] == 1:
+                    index = np.arange(height.shape[0])
+                    if bottom is None:
+                        bottom = np.zeros(height.shape)
+
+                else:
+                    raise ValueError('height should only have one column')
+
+            else:
+                raise ValueError('3d is not expected, just 2d')
+        else:
+            index = np.arange(len(height))
+    else:
+        ValueError("Unknown ValueError: height")
+
+    return index, height, bottom
+
+
+def calc_vertices(index, height, bottom, width):
+    # construct our vertices
+
+    X, Y = (0, 1)
+
+    rr = np.zeros((6 * len(index), 3), np.float32)
+    rr[:, X] = np.repeat(index - 0.5, 6)
+    rr[0::6, X] += (1 - width) / 2
+    rr[1::6, X] += (1 - width) / 2
+    rr[2::6, X] += 1 - (1 - width) / 2
+    rr[3::6, X] += (1 - width) / 2
+    rr[4::6, X] += 1 - (1 - width) / 2
+    rr[5::6, X] += 1 - (1 - width) / 2
+    rr[:, Y] = np.repeat(bottom, 6)
+    rr[::2, Y] += np.repeat(height, 3)
+    index.astype(np.float32)
+
+    return rr
```

Comment:
I noticed that calling `Bar.meshdata.get_vertices()` errors out. It turns out that `mesh` only half-works if you give it a `(n, 3)` array of vertices without faces (as it actually expects a `(n, 3, 3)` array of explicit faces by coordinates). For some reason it *mostly* goes silently, but some things clearly don't work.

The easiest way to solve it, I think, is to reshape the return of this function:

```
return rr.reshape(-1, 3, 3)
```

Alternatively, go the [`Histogram`](https://github.com/vispy/vispy/blob/main/vispy/visuals/histogram.py) approach and explicitly pass an array to indicate the faces.